README.md
# polld 
[![NPM version][npm-image]][npm-url] [![Dependency Status][daviddm-url]][daviddm-image]

Because sometimes, it's just easier to poll.

## Install

```bash
$ npm install -g polld
```

## Usage

```bash
polld -h <statsd_host> -p <statsd_port> -i <interval> <taskfile>
```
The taskfile is just a .js file that is run at each interval period. Provided inside the taskfile is a handle to an instance of [node-statsd](https://github.com/sivy/node-statsd) to send metrics to the statsd server.

The taskfile can be written in js or coffee, See /example for more details.

## License

Copyright (c) 2015 Ainsley Chong. Licensed under the MIT license.


[npm-url]: https://npmjs.org/package/polld
[npm-image]: https://badge.fury.io/js/polld.svg
[travis-url]: https://travis-ci.org/ainsleyc/polld
[travis-image]: https://travis-ci.org/ainsleyc/polld.svg?branch=master
[daviddm-url]: https://david-dm.org/ainsleyc/polld.svg?theme=shields.io
[daviddm-image]: https://david-dm.org/ainsleyc/polld
[coveralls-url]: https://coveralls.io/r/ainsleyc/polld
[coveralls-image]: https://coveralls.io/repos/ainsleyc/polld/badge.png


CHANGELOG.md

# 0.1.0 (2015-05-27)

- Initial version


test/options.js
'use strict';

var should = require('should');
var _ = require('lodash');

var options = require('../lib/options');

function parameterCheck(params) {
  it('should have a label', function() {
    should.exist(params.label);
  });

  it('should have a symbol', function() {
    should.exist(params.sym);
  });

  it('should have a required', function() {
    should.exist(params.required);
  });

  it('should have a validator', function() {
    should.exist(params.required);
  });

  it('should have a default if not required', function() {
    if(!params.required) {
      should.exist(params.default);
    }
  });
}

function positiveIntegerCheck(field) {
  it('should not allow non-integer values', function() {
    options[field].validator('blah').should.be.false
  });

  it('should not allow negative integers', function() {
    options[field].validator(-1).should.be.false
  });

  it('should allow positive integers', function() {
    options[field].validator(100).should.be.true
  });
}

describe('options', function () {

  describe('host', function() {
    parameterCheck(options.host);
  });

  describe('port', function() {
    parameterCheck(options.port);
    positiveIntegerCheck('port');
  });

  describe('interval', function() {
    parameterCheck(options.interval);
    positiveIntegerCheck('interval');
  });

});


test/run_test.js
'use strict';

var assert = require('should');
var sinon = require('sinon');

var options = require('../lib/options');
var taskRunner = require('../lib/taskRunner');

var statsdStub = {};
function emptyTask(statsd, cb) {
  cb(null);
}

describe('taskRunner', function () {

  it('should call the task iteration number of times', function (done) {
    var spy = sinon.spy(emptyTask);
    taskRunner._startIntervals(spy, statsdStub, 10, 10, function(err, count) {
      spy.callCount.should.be.equal(10);
      done();
    });
  });

  it('should return number of iterations', function (done) {
    var spy = sinon.spy(emptyTask);
    taskRunner._startIntervals(emptyTask, statsdStub, 10, 10, function(err, count) {
      count.should.be.equal(10);
      done();
    });
  });

});


test/polld_test.js
'use strict';

var child_process = require('child_process');
var should = require('should');

function checkFailedExit(cmd) {
  (function() { child_process.execSync(cmd) }).should.throw(Error);
  try {
    child_process.execSync(cmd);
  } catch (e) {
    e.status.should.equal(1);
  }
}

describe('polld', function () {

  it('should fail for invalid port', function () {
    checkFailedExit('./bin/polld -p blah ./example/task.js');
  });

  it('should fail for invalid interval', function () {
    checkFailedExit('./bin/polld -i blah ./example/task.js');
  });

  it('should fail for missing taskfile', function () {
    checkFailedExit('./bin/polld ../example/doesNotExist.js');
  });

  it('should fail for missing "task" function in taskfile', function () {
    checkFailedExit('./bin/polld ./package.json');
  });

});


gulpfile.js
'use strict';

var gulp   = require('gulp');
var plugins = require('gulp-load-plugins')();

var paths = {
  lint: ['./gulpfile.js', './lib/**/*.js'],
  tests: ['./test/**/*.js', '!test/{temp,temp/**}'],
  watch: [
    './gulpfile.js',
    './lib/**',
    './test/**/*.js',
    '!test/{temp,temp/**}'
  ],
};

var plumberConf = {};

if (process.env.CI) {
  plumberConf.errorHandler = function(err) {
    throw err;
  };
}

gulp.task('lint', function () {
  return gulp.src(paths.lint)
    .pipe(plugins.jshint('.jshintrc'))
    .pipe(plugins.plumber(plumberConf))
    .pipe(plugins.jscs())
    .pipe(plugins.jshint.reporter('jshint-stylish'));
});

gulp.task('unitTest', function () {
  gulp.src(paths.tests, {cwd: __dirname})
    .pipe(plugins.plumber(plumberConf))
    .pipe(plugins.mocha({ reporter: 'list' }));
});

gulp.task('watch', ['test'], function () {
  gulp.watch(paths.watch, ['test']);
});

gulp.task('test', ['lint', 'unitTest']);

gulp.task('default', ['test']);


lib/options.js
'use strict';

var validate = require('validate.js');

var options = {
  host: {
    label: 'statsd host',
    sym: 'h',
    default: 'localhost',
    required: false,
    validate: function (value) {
      if (!validate.isString(value)) { return false; }
      return true;
    }
  },

  port: {
    label: 'statsd port',
    sym: 'p',
    default: 8125,
    required: false,
    validate: function (value) {
      var intValue = parseInt(value);
      if (!validate.isInteger(intValue)) { return false; }
      if (intValue <= 0) { return false; }
      return true;
    }
  },

  interval: {
    label: 'reporting interval in ms',
    sym: 'i',
    default: 10000,
    required: false,
    validate: function (value) {
      var intValue = parseInt(value);
      if (!validate.isInteger(intValue)) { return false; }
      if (intValue <= 0) { return false; }
      return true;
    }
  }
};

module.exports = options;


lib/taskRunner.js
'use strict';

require('coffee-script/register');
var StatsD = require('node-statsd');

function initStatsd(options) {
  var statsdOpts = {
    host: options.host,
    port: options.port,
    prefix: options.prefix || '',
    suffix: options.suffix || ''
  };
  return new StatsD(statsdOpts);
}

function loadTaskFn(taskfile) {
  return require(taskfile);
}

function startIntervals(taskFn, statsd, interval, iterations, cb) {
  var finishedIntervals = 0;
  var intervalHandle = setInterval(function() {
    taskFn(statsd, function(err) {
      if (err) {
        console.log(err);
      }
    });
    finishedIntervals++;
    if (finishedIntervals >= iterations) {
      clearInterval(intervalHandle);
      if (cb) {
        cb(null, finishedIntervals);
      }
    }
  }, interval);
}

function run(taskfile, options, cb) {

  var taskFn = loadTaskFn(taskfile).task;
  var statsd = initStatsd(options);

  var interval = options.interval;
  var iterations = options.iterations;

  startIntervals(taskFn, statsd, interval, iterations, function(err) {
    cb(err);
  });
}

var taskRunner = {
  _startIntervals: startIntervals,
  run: run
};

module.exports = taskRunner;


lib/polld.js
/*
 * polld
 * https://github.com/ainsleyc/polld.git
 *
 * Copyright (c) 2015 Ainsley Chong
 * Licensed under the MIT license.
 */

'use strict';

var _ = require('lodash');
var program = require('commander');

var options = require('./options');
var taskRunner = require('./taskRunner');

function failureExit(message) {
  console.log(message);
  process.exit(1);
}

function getTaskPath(taskfile) {
  return [process.cwd(), '/', taskfile].join('');
}

function invalidParamMessage(param, value) {
  return ['Error: "', value, '" is an invalid ', param, ' parameter'].join('');
}

function validateParams(taskfile, opts) {
  if (!options.port.validate(opts.port)) {
    failureExit(invalidParamMessage('port', opts.port));
  }

  if (!options.port.validate(opts.interval)) {
    failureExit(invalidParamMessage('interval', opts.interval));
  }

  var taskObject = null;
  try {
    taskObject = require(getTaskPath(taskfile));
  } catch (e) {
    failureExit(invalidParamMessage('taskfile', taskfile));
  }

  if (!_.isFunction(taskObject.task)) {
    failureExit('taskfile does not correctly implement a "task" function');
  }
}

function start(taskfile, opts) {
  validateParams(taskfile, opts);
  taskRunner.run(getTaskPath(taskfile), opts);
}

program
  .version('0.0.1')
  .usage('[options] <taskfile>')
  .option('-h, --host [host]', 'statsd host [localhost]', 'localhost')
  .option('-p, --port [port]', 'statsd port [8125]', 8125)
  .option('-i, --interval [ms]', 'reporting interval in ms [10000]', 10000)
  .option('-r, --prefix', 'statsd prefix')
  .option('-s, --suffix', 'statsd suffix')
  .action(start)
  .parse(process.argv);

